### ISSUE TITLE
torch.Tensor returns False for PySequence_Check and True for PyNumber_Check

### ISSUE THREAD

**Issue by mroeschke** [pytorch/pytorch#99646]
### 🐛 Describe the bug

```
In [9]: %%cython
   ...: from cpython.sequence cimport PySequence_Check
   ...: def f(x):
   ...:     return PySequence_Check(x)
   ...: 
ld: warning: -pie being ignored. It is only used when linking a main executable

In [10]: import torch

In [11]: f(torch.tensor([1, 2, 3]))
Out[11]: False  # expect True

In [12]: %%cython
    ...: from cpython.number cimport PyNumber_Check
    ...: def f(x):
    ...:     return PyNumber_Check(x)
    ...: 
ld: warning: -pie being ignored. It is only used when linking a main executable

In [13]: f(torch.tensor([1, 2, 3]))
Out[13]: True  # expect False

In [14]: torch.__version__
Out[14]: '2.0.0'
```

### Versions

```
% conda list | grep torch
pytorch                   2.0.0                   py3.8_0    pytorch
torchaudio                2.0.0                  py38_cpu    pytorch
torchvision               0.13.1          cpu_py38he5d19f7_0  
```

cc @albanD

**Comment by soulitzer** [comment]
According to https://docs.python.org/3/c-api/number.html, an object should pass the PyNumber_Check if it implements the number protocol, so maybe it makes sense that it is a "number.

I don't know if tensors necessarily need to pass PySequence_Check either since 0-dim tensors also exist which don't really have the indexing operation defined on them.

**Comment by albanD** [comment]
Hey!
This is indeed the current behavior, mostly because we do not implement the tp_sequence protocol.
I'm not sure why PyNumber_Check is true though...

A lot of our C binding code does rely on these though so I'm not sure if we can change this.

**Comment by mroeschke** [comment]
Okay thanks for the response. I guess since this is intended behavior I'll close.
